agroal/src/main/java/org/wildfly/extension/agroal/operation/XADataSourceOperations.java

```diff
@@ -0,0 +1,122 @@
+
+    public static final OperationStepHandler REMOVE_OPERATION = new XADataSourceRemove();
+
+    // --- //
+
+    private static class XADataSourceAdd extends AbstractAddStepHandler {
+
+        @Override
+        protected void populateModel(ModelNode operation, ModelNode model) throws OperationFailedException {
+            for ( AttributeDefinition attributeDefinition : XADataSourceDefinition.INSTANCE.getAttributes() ) {
+                attributeDefinition.validateAndSet( operation, model );
+            }
+        }
+
+        @Override
+        protected void performRuntime(OperationContext context, ModelNode operation, ModelNode model) throws OperationFailedException {
+            String datasourceName = PathAddress.pathAddress( operation.require( OP_ADDR ) ).getLastElement().getValue();
+
+            ModelNode factoryModel = CONNECTION_FACTORY_ATTRIBUTE.resolveModelAttribute( context, model );
+            AgroalConnectionFactoryConfigurationSupplier connectionFactoryConfiguration = AbstractDataSourceOperations.connectionFactoryConfiguration( context, factoryModel );
+
+            ModelNode poolModel = CONNECTION_POOL_ATTRIBUTE.resolveModelAttribute( context, model );
+            AgroalConnectionPoolConfigurationSupplier connectionPoolConfiguration = AbstractDataSourceOperations.connectionPoolConfiguration( context, poolModel );
+            connectionPoolConfiguration.connectionFactoryConfiguration( connectionFactoryConfiguration );
+
+            AgroalDataSourceConfigurationSupplier dataSourceConfiguration = new AgroalDataSourceConfigurationSupplier();
+            dataSourceConfiguration.connectionPoolConfiguration( connectionPoolConfiguration );
+            dataSourceConfiguration.metricsEnabled( STATISTICS_ENABLED_ATTRIBUTE.resolveModelAttribute( context, model ).asBoolean() );
+
+            String jndiName = JNDI_NAME_ATTRIBUTE.resolveModelAttribute( context, model ).asString();
+
+            String driverName = DRIVER_ATTRIBUTE.resolveModelAttribute( context, factoryModel ).asString();
+            ServiceName driverServiceName = ServiceName.of( DriverOperations.DRIVER_SERVICE_PREFIX, driverName );
+
+            DataSourceService dataSourceService = new DataSourceService( datasourceName, jndiName, false, true, dataSourceConfiguration );
+            ServiceName dataSourceServiceName = XADATASOURCE_SERVICE_PREFIX.append( datasourceName );
+
+            ServiceBuilder<AgroalDataSource> serviceBuilder = context.getServiceTarget().addService( dataSourceServiceName, dataSourceService );
+            serviceBuilder.addDependency( driverServiceName, DriverService.ProviderClass.class, dataSourceService.getDriverServiceInjector() );
+
+            // Define JTA dependencies (required for XA)
+            serviceBuilder.addDependency( TxnServices.JBOSS_TXN_TRANSACTION_MANAGER, TransactionManager.class, dataSourceService.getTransactionManagerInjector() );
```

Comment:
the ServiceBuilder API is fluent, you can chain the calls to `addDependency(...)`